### PR TITLE
1. Add Currency and Currency Rate type Field. 

### PR DESCRIPTION
**Pull request by KotiGungi-Vienna**
1. Add Currency and Currency Rate type Field. 
2. At the time of Multiple Invoice with different currency (from the bank currency), System will be set Currency as on invoice  but User can change the Currency  between bank or Invoice currency and System will set currency rate type with Latest Invoice Currency Rate type (it should be editable, user as the choice to change). Payment will created in the currency and currency rate type set on the Line. 
Note:  User can only drag the Same currency invoice. 
Example: Invoice-USD drag to new form, now user cannot drag the AED Invoice. 
3. At the time of Order with different currency, System will be set Currency as per Order but User can change the Currency  between bank or order currency and System will set currency rate type bind on the Order Currency Rate type (it should be editable, user as the choice to change). Payment will created in the currency and currency rate type set on the Line. 
Note:  User cannot reconcile the multiple order. 

4. At the time of Order with same currency (from bank currency), System will be set Currency same on new form and System will set currency rate type bind on the Order Currency Rate type (it should be editable, user as the choice to change). Payment will created in the currency and currency rate type set on the Line. 

5. At the time of Invoice  same currency (from bank currency) (from the bank currency) System will be set Currency same on new form and System will set currency rate type with Latest Invoice Currency Rate type (it should be editable, user as the choice to change). Payment will created in the currency and currency rate type set on the Line. 

6. At the time of  Payment with different currency, System will be set Currency as on bank account and System will set currency rate type same on payment window. 
Note: User can only drag the one payment at a time 

7. At the time of Payment with Same currency  (from the bank currency), System will be set Currency same and  Currency rate type will be set same as on the Payment window.  

8.  Charge is Present without any Payment or Invoice and Order then Currency Rate Type same on the transaction and currency will be same as on transaction. 
Note: User can only drag the one payment at a time

### CODE REVIEW

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 1]
```diff
@@ -743,84 +751,51 @@ FROM C_PERIOD
...
-                    if (string.IsNullOrEmpty(error))
+                MBankStatement _bankStatement = null;
+                /*change by pratap*/
+                MPayment paymentrecord = null;
+                MDocType documentType = null;
+                /*change by pratap*/
+                MBankStatementLine _bankStatementLine = null;
+                if (_existingStatementID <= 0)
+                {
+                    _bankStatement = new MBankStatement(ctx, 0, null);
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:762]: use transaction

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 2]
```diff
@@ -743,84 +751,51 @@ FROM C_PERIOD
...
 
-            if (_formData[0]._bankStatementLineID > 0)
-            {
-                _bankStatementLine = new MBankStatementLine(ctx, _formData[0]._bankStatementLineID, null);
-            }
-            else
-            {
-                _bankStatementLine = new MBankStatementLine(_bankStatement);
-            }
+                    _bankStatement = new MBankStatement(ctx, _existingStatementID, null);
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:789]: use trx

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 3]
```diff
@@ -1171,9 +1136,12 @@ FROM C_PERIOD
...
                     _bankStatementLine.SetC_BPartner_ID(0);
                 }
-                if (_formData[0]._ctrlInvoice > 0)
+                //_formData[0]._ctrlInvoice - Type is changed int to string so according to that check the value using IsNullOrEmpty() method
+                if (!string.IsNullOrEmpty(_formData[0]._ctrlInvoice))
                 {
-                    _bankStatementLine.SetC_Invoice_ID(_formData[0]._ctrlInvoice);
+                    //If Invoice is have then set Invoice from Payment
+                    //_bankStatementLine.SetC_Invoice_ID(_formData[0]._ctrlInvoice);
+                    _bankStatementLine.SetC_Invoice_ID(paymentrecord.GetC_Invoice_ID());
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:1144]: handle null execption when payement record is null..

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 4]
```diff
@@ -1276,29 +1258,194 @@ INTO VA012_TEMPSTATEMENT
...
+        /// <returns>List</returns>
+        public Dictionary<String, Object> GetConvertedAmount(Ctx ctx, int currency, int conversionType, DateTime? stmtDate, string schedules, int _accountId, int orderId)
+        {
+            decimal _convertedAmt = 0;
+            DataSet _ds = null;
+            string[] ids = null;
+            string _query = null;
+            Dictionary<String, Object> _list = new Dictionary<String, Object>();
+
+            int _org_ID = Util.GetValueOfInt(DB.ExecuteScalar("SELECT C_CURRENCY_ID, AD_Org_ID FROM C_BANKACCOUNT WHERE C_BANKACCOUNT_ID=" + _accountId, null, null));
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:1304]: correct this query

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 5]
```diff
@@ -1276,29 +1258,194 @@ INTO VA012_TEMPSTATEMENT
...
+            //    " pay.IsActive = 'Y' AND C_InvoicePaySchedule_ID IN (" + schedules + ")";
+            if (!string.IsNullOrEmpty(schedules))
+            {
+                _query = @"SELECT CASE
+                                    WHEN(inv.C_CURRENCY_ID != BCURR.C_CURRENCY_ID)
+                                    THEN
+                                    CASE
+                                     WHEN(DT.DOCBASETYPE = 'ARI' OR DT.DOCBASETYPE = 'APC')
+                                     THEN CURRENCYCONVERT(PAY.DueAmt, inv.C_CURRENCY_ID, BCURR.C_CURRENCY_ID, " + GlobalVariable.TO_DATE(stmtDate, true) + "," + conversionType + @", INV.AD_Client_ID, " + _org_ID + @")
+                                     ELSE CURRENCYCONVERT(PAY.DueAmt*-1, inv.C_CURRENCY_ID, 304, " + GlobalVariable.TO_DATE(stmtDate, true) + @"
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:1316]: dont fix currency

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 6]
```diff
@@ -1276,29 +1258,194 @@ INTO VA012_TEMPSTATEMENT
...
+            }
+            return convsionTypeList;
+        }
+
+        /// <summary>
+        /// Get Currency list
+        /// </summary>
+        /// <param name="ctx">Context</param>
+        /// <returns>List</returns>
+        public List<MatchBase> GetCurrency(Ctx ctx)
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:1433]: use control

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 7]
```diff
@@ -4095,6 +4215,9 @@ public string CreatePaymentFromSchedule(Ctx ctx, List<StatementProp> _formData)
                                     return !string.IsNullOrEmpty(error) ? error : "VA012_PaymentNotSaved";
                                 }
                             }
+                            //fetch the Updated data from new Instance 
+                            _pay = new MPayment(ctx, _pay.GetC_Payment_ID(), null);
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:4219]: implement Complete Action call by worker, not to call direct

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 8]
```diff
@@ -4251,8 +4379,16 @@ public string CreatePaymentFromOrder(Ctx ctx, List<StatementProp> _formData)
...
+                {
+                    _txtAmount = MConversionRate.Convert(ctx, _formData[0]._txtAmount, _formData[0]._cmbCurrency, _formData[0]._txtCurrency, _formData[0]._dtStatementDate, _formData[0]._txtConversionType, ctx.GetAD_Client_ID(), _formData[0]._bankAcctOrg_ID);
+                }
+                else {
+                    _txtAmount = _formData[0]._txtAmount;
+                }
+
                 MPayment _pay = new MPayment(ctx, 0, null);
-                int C_Doctype_ID = GetDocTypeID(ctx, _formData[0]._txtAmount);
+                int C_Doctype_ID = GetDocTypeID(ctx, _txtAmount);//Get DocType using Amount
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:4391]: correct this method, dont pick organization from the context

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 9]
```diff
@@ -4920,6 +5086,22 @@ ELSE CCR1.DIVIDERATE
                         //}
                         //end pratap
                     }
+                    //used new query to get the C_Currency_ID
+                    _sql = @"SELECT UNIQUE INV.C_Currency_ID
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:5090]: make it Distinct
Unique will not work in postgreSQL

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 10]
```diff
@@ -5102,13 +5296,20 @@ public ContraResponse CheckContraCondition(Ctx ctx, int _dragSourceID, int _drag
...
                                         ON " + _currencyId + @" =BCURR.C_CURRENCY_ID
                                         WHERE csl.C_cashline_ID=" + _dragSourceID;
-                _payAmt = Decimal.Negate(Util.GetValueOfDecimal(DB.ExecuteScalar(_sql)));
+                DataSet _ds = DB.ExecuteDataset(_sql, null, null);
+                //_payAmt = Decimal.Negate(Util.GetValueOfDecimal(DB.ExecuteScalar(_sql)));
+                //get the payAmt as well C_Currency_ID
+                if (_ds != null && _ds.Tables[0].Rows.Count > 0)
+                {
+                    _payAmt = Decimal.Negate(Util.GetValueOfDecimal(_ds.Tables[0].Rows[0]["AMOUNT"]));
+                    _currency_Id = Util.GetValueOfInt(_ds.Tables[0].Rows[0]["AMOUNT"]);
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:5311]: change the columnname

File: ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js [hunk 11]
```diff
@@ -27,6 +27,31 @@
...
+    VA012_Contra.prototype.SetConversionType = function (ctx, windowNo, mTab, mField, value, oldValue) {
+        if (this.isCalloutActive()) {
+            return "";
+        }
+        this.setCalloutActive(true);
+        if (value == null) {
+            return "";
+        }
+        else {
+            if (VIS.Utility.Util.getValueOfInt(value) > 0) {
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js:41]: handle exception if occure

File: ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js [hunk 12]
```diff
@@ -27,6 +27,31 @@
         return "";
 
     };
+
+    //Callout to SetConversionType on statement Line
+    VA012_Contra.prototype.SetConversionType = function (ctx, windowNo, mTab, mField, value, oldValue) {
+        if (this.isCalloutActive()) {
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js:33]: take this function before callout mark as active

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js [hunk 13]
```diff
@@ -2114,15 +2179,27 @@
...
+                                        if (VIS.Utility.Util.getValueOfInt(_cashLineSelectedVal) == VIS.Utility.Util.getValueOfInt(($(ui.draggable)).data('uid'))) {
+                                            VIS.ADialog.info("VA012_AlreadySelected", null, "", "");
+                                        }
+                                        else {
+                                            VIS.ADialog.info("VA012_PlzRemoveOrUndoPrevSelectedRecord", null, "", "");
+                                        }
+                                        return;
+                                    }
+                                    else {
+                                        if (loadFunctions.checkContraCondition(($(ui.draggable)).data('uid'), $(this).attr("data-uid"), _txtAmount.getValue())) {
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js:2193]: check this condition with execution

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js [hunk 14]
```diff
@@ -4227,9 +4453,11 @@
                         _btnIn.attr("v_active", "0");
                     }
                     _txtAmount.getControl().trigger('blur');
+                    //Set the Currency and conversion Type
+                    setCurrencyandConversionType(_scheduleList.toString());
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js:4457]: check schedule list having value in all case or will be null then handle exception

**Review by Amit-Vienna — CHANGES_REQUESTED** [review]
do the changes as suggested

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 1 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 3 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 4 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 5 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 6 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 7 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 8 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 9 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 10 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs]: Done.

File: ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js (hunk 11 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js:42]: Done.

File: ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js (hunk 12 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Scripts/model/callouts.js]: Done.

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js (hunk 13 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js:2192]: Done.

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js (hunk 14 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js]: Done.

**Comment by KotiGungi-Vienna** [comment]
done all changes.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 15]
```diff
@@ -3815,11 +3910,18 @@ public int GetDocTypeID(Ctx ctx, decimal _amount)
             {
                 _docBaseType = "APP";
             }
-            return Util.GetValueOfInt(DB.ExecuteScalar("Select  dt.c_doctype_id  From C_doctype DT inner join c_docbasetype DBT On dt.docbasetype=dbt.docbasetype where dbt.docbasetype='" + _docBaseType + "' AND dt.IsActive = 'Y' AND (DT.ad_org_id = " + ctx.GetAD_Org_ID() + " or  DT.ad_org_id = 0) AND DT.AD_Client_ID = " + ctx.GetAD_Client_ID()));
+            return Util.GetValueOfInt(DB.ExecuteScalar("Select  dt.c_doctype_id  From C_doctype DT inner join c_docbasetype DBT On dt.docbasetype=dbt.docbasetype where dbt.docbasetype='" + _docBaseType + "' AND dt.IsActive = 'Y' AND (DT.ad_org_id = " + ctx.GetAD_Org_ID() + " or  DT.ad_org_id = 0) AND DT.AD_Client_ID = " + org_Id));
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:3913]: change variable

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 16]
```diff
@@ -3829,7 +3931,7 @@ public int GetDocTypeID(Ctx ctx, string docBaseType)
             {
                 docBaseType = "ARR";
             }
-            return Util.GetValueOfInt(DB.ExecuteScalar("Select  MIN(dt.c_doctype_id)  From C_doctype DT inner join c_docbasetype DBT On dt.docbasetype=dbt.docbasetype where dbt.docbasetype='" + docBaseType + "' AND dt.IsActive = 'Y' AND (DT.ad_org_id = " + ctx.GetAD_Org_ID() + " or  DT.ad_org_id = 0) AND DT.AD_Client_ID = " + ctx.GetAD_Client_ID()));
+            return Util.GetValueOfInt(DB.ExecuteScalar("Select  MIN(dt.c_doctype_id)  From C_doctype DT inner join c_docbasetype DBT On dt.docbasetype=dbt.docbasetype where dbt.docbasetype='" + docBaseType + "' AND dt.IsActive = 'Y' AND (DT.ad_org_id = " + ctx.GetAD_Org_ID() + " or  DT.ad_org_id = 0) AND DT.AD_Client_ID = " + org_Id));
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:3934]: change passed variable into query

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js [hunk 17]
```diff
@@ -6732,56 +6759,40 @@
...
-                        }
-                    }
+            loadNewFormCurrency: function () {
+                var getCurrency = null;
+                //clear the previous options
+                $(_txtCurrency[0]).empty();
+                //shown the records only IsMycurrency is true.
+                var _txtCurrencyLookUp = VIS.MLookupFactory.get(VIS.Env.getCtx(), $self.windowNo, 457, VIS.DisplayType.TableDir, "C_Currency_ID", 0, false, "IsMyCurrency='Y'");
+                var getCurrency = _txtCurrencyLookUp.getData(true, true, false, false);
+                var flag = 0;
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js:6766]: if not required, then removed it

**Review by Amit-Vienna — CHANGES_REQUESTED** [review]
do the changes as needful

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 15 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:3913]: Done.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs (hunk 16 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:3934]: Done.

File: ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js (hunk 17 above)
Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Scripts/apps/forms/bankstatement.js:6766]: Done.

**Comment by KotiGungi-Vienna** [comment]
Done all.

File: ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs [hunk 18]
```diff
@@ -677,11 +642,15 @@ FROM C_PERIOD
...
                             {
+                                //closing transaction
+                                trx.Rollback();
                                 return "VA012_StatementAlreadyExistDiffAcc";
                             }
                         }
                         else if (_statementDocStatus == "CO" || _statementDocStatus == "CL" || _statementDocStatus == "RE" || _statementDocStatus == "VO")
                         {
+                            //closing transaction
+                            trx.Rollback();
```
Comment (Amit-Vienna) [inline ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:653]: before return, close the self created transaction

Comment (KotiGungi-Vienna) [reply ViennaAdvantage/Areas/VA012/Models/BankStatementModel.cs:661]: Done.

**Review by Amit-Vienna — APPROVED** [review]
fine

**Pull request merged**